### PR TITLE
[Promotion] Coupon validation

### LINKED ISSUE #5008: [Promotion] Flash message shown in every step of checkout
When I checkout with a promotion, I get a success flash (sometimes multiple) in each of the checkout steps.

### PR DESCRIPTION
**Pull request by pamil**
| Q               | A
| --------------- | ---
| Bug fix?        | yes
| New feature?    | yes
| BC breaks?      | no
| Related tickets | fixes #5008, #5085
| License         | MIT

### CODE REVIEW

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php [hunk 1]
```diff
@@ -164,15 +164,15 @@ public function myEmailShouldBe($email)
     }
 
     /**
-     * @Then /^I should be notified that the ([^"]+) is required$/
+     * @Then /^I should be notified that the (email|password|city|street|first name|last name) is required$/
```
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: Is it really connected to this PR?

File: src/Sylius/Bundle/PromotionBundle/Form/Type/CouponToCodeType.php [hunk 2]
```diff
@@ -56,13 +56,50 @@ public function configureOptions(OptionsResolver $resolver)
             ->setDefaults([
                 'data_class' => null,
                 'label' => 'sylius.ui.code',
+                'validation_groups' => function (FormInterface $form) {
+                    $groups = ['sylius']; // Regular validation groups
+
+//                    if ((bool) $form->getData()) { // Validate the coupon if it was sent
```
Comment (lchrusciel) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/CouponToCodeType.php]: 🚜

Comment (pamil) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/CouponToCodeType.php]: Good catch! :D

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml [hunk 3]
```diff
@@ -124,6 +124,17 @@
             </constraint>
         </property>
     </class>
+
```
Comment (ibasaw) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml]: I saw that there are 2 same classname for Sylius\Component\Core\Model\Order in the file

Already exist at line 61

Is it the problem for failed test ?

File: features/promotion/receiving_discount/receiving_fixed_discount_on_cart_from_coupon.feature [hunk 4]
```diff
@@ -9,19 +9,46 @@ Feature: Receiving fixed discount on cart from coupon
...
     Scenario: Receiving no discount from invalid coupon
-        Given this promotion gives "$10.00" discount to every order
         When I add product "PHP T-Shirt" to the cart
         And I use coupon with code "SANTA2011"
+        Then I should be notified that the promotion coupon is invalid
+        And my cart total should be "$100.00"
+        And there should be no discount
+
+    @ui
+    Scenario: Receiving no discount from expired coupon
```
Comment (lchrusciel) [inline features/promotion/receiving_discount/receiving_fixed_discount_on_cart_from_coupon.feature]: We should move this scenarios to separate feature file

File: src/Sylius/Behat/Context/Ui/Shop/AccountContext.php (hunk 1 above)
Comment (lchrusciel) [inline src/Sylius/Behat/Context/Ui/Shop/AccountContext.php]: One more time. Shouldn't it be reverted?

File: src/Sylius/Bundle/PromotionBundle/Validator/PromotionCouponValidator.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+        if (!$value instanceof Coupon) {
+            $this->context->addViolation($constraint->invalidMessage);
+
+            return;
+        }
+
+        if ($value->getExpiresAt() !== null && $value->getExpiresAt() < new \DateTime()) {
+            $this->context->addViolation($constraint->expiredMessage);
+
+            return;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/PromotionBundle/Validator/PromotionCouponValidator.php]: Should we really serve only one message? Another thing is that, this and the statement below could be coupled.

File: features/promotion/receiving_discount/receiving_fixed_discount_on_cart_from_coupon.feature [hunk 6]
```diff
@@ -9,19 +9,46 @@ Feature: Receiving fixed discount on cart from coupon
...
+    Scenario: Receiving no discount from valid coupon from expired promotion
+        Given this promotion has already expired
+        When I add product "PHP T-Shirt" to the cart
+        And I use coupon with code "SANTA2016"
+        Then I should be notified that the promotion coupon has expired
+        And my cart total should be "$100.00"
+        And there should be no discount
+
+    @ui
+    Scenario: Receiving no discount from valid coupon that has reached its usage limit
```
Comment (GSadee) [inline features/promotion/receiving_discount/receiving_fixed_discount_on_cart_from_coupon.feature]: I guess that coupon that has reached its usage limit is not valid anymore

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 7]
```diff
@@ -130,11 +130,44 @@ public function thereIsPromotionWithCoupon($promotionName, $couponCode, $usageLi
         $promotion->setCouponBased(true);
 
         $this->promotionRepository->add($promotion);
+
         $this->sharedStorage->set('promotion', $promotion);
+        $this->sharedStorage->set('promotion_coupon', $coupon);
         $this->sharedStorage->set('coupon', $coupon);
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: Is that necessary to setting the same coupon with two different keys?

File: src/Sylius/Behat/Context/Setup/PromotionContext.php [hunk 8]
```diff
@@ -130,11 +130,44 @@ public function thereIsPromotionWithCoupon($promotionName, $couponCode, $usageLi
...
+    {
+        $coupon->setExpiresAt(new \DateTime('1 day ago'));
+
+        $this->objectManager->flush();
+    }
+
+    /**
+     * @Given /^(this promotion coupon) has already reached its usage limit$/
+     */
+    public function thisPromotionCouponHasReachedItsUsageLimit(CouponInterface $coupon)
```
Comment (GSadee) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: I guess that you can merge this method with: https://github.com/Sylius/Sylius/blob/master/src/Sylius/Behat/Context/Setup/PromotionContext.php#L460

File: src/Sylius/Behat/Context/Setup/PromotionContext.php (hunk 7 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: It's either a promotion coupon or just a coupon, need to unify this later.

File: src/Sylius/Behat/Context/Setup/PromotionContext.php (hunk 8 above)
Comment (pamil) [inline src/Sylius/Behat/Context/Setup/PromotionContext.php]: :+1:

File: features/promotion/receiving_discount/receiving_fixed_discount_on_cart_from_coupon.feature (hunk 4 above)
Comment (pamil) [inline features/promotion/receiving_discount/receiving_fixed_discount_on_cart_from_coupon.feature]: Done.

File: src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml (hunk 3 above)
Comment (pamil) [inline src/Sylius/Bundle/CoreBundle/Resources/config/validation.xml]: Thanks, fixed!

**Comment by pamil** [comment]
Closing in favour of #6273.

**Pull request closed without merging**